rust/arrow/src/array/builder.rs

```diff
@@ -973,6 +995,105 @@ impl Drop for StructBuilder {
     }
 }
 
+/// Array builder for `DictionaryArray`. For example to map a set of byte indices
+/// to f32 values. Note that the use of a `HashMap` here will not scale to very large
+/// arrays or result in an ordered dictionary.
+pub struct PrimitiveDictionaryBuilder<K, V>
+where
+    K: ArrowPrimitiveType,
+    V: ArrowPrimitiveType,
+{
+    keys_builder: PrimitiveBuilder<K>,
+    values_builder: PrimitiveBuilder<V>,
+    map: HashMap<&'static [u8], K::Native>,
+}
+
+impl<K, V> PrimitiveDictionaryBuilder<K, V>
+where
+    K: ArrowPrimitiveType,
+    V: ArrowPrimitiveType,
+{
+    /// Creates a new `PrimitiveDictionaryBuilder` from a keys builder and a value builder.
+    pub fn new(
+        keys_builder: PrimitiveBuilder<K>,
+        values_builder: PrimitiveBuilder<V>,
+    ) -> Self {
+        Self {
+            keys_builder: keys_builder,
+            values_builder: values_builder,
+            map: HashMap::new(),
+        }
+    }
+}
+
+impl<K, V> ArrayBuilder for PrimitiveDictionaryBuilder<K, V>
+where
+    K: ArrowPrimitiveType,
+    V: ArrowPrimitiveType,
+{
+    /// Returns the builder as an non-mutable `Any` reference.
+    fn as_any(&self) -> &Any {
+        self
+    }
+
+    /// Returns the builder as an mutable `Any` reference.
+    fn as_any_mut(&mut self) -> &mut Any {
+        self
+    }
+
+    /// Returns the boxed builder as a box of `Any`.
+    fn into_box_any(self: Box<Self>) -> Box<Any> {
+        self
+    }
+
+    /// Returns the number of array slots in the builder
+    fn len(&self) -> usize {
+        self.keys_builder.len()
+    }
+
+    /// Builds the array and reset this builder.
+    fn finish(&mut self) -> ArrayRef {
+        Arc::new(self.finish())
+    }
+}
+
+impl<K, V> PrimitiveDictionaryBuilder<K, V>
+where
+    K: ArrowPrimitiveType,
+    V: ArrowPrimitiveType,
+{
+    /// Append a primive value to the array. Return an existing index
+    /// if already present in the values array or a new index if the
+    /// value is appended to the values array.
+    /// ```
```

Comment:
I was planning to add an example here. It didn't make it to the changelist.

Reply:
It is difficult (impossible?) to return the keys as an ArrayRef as the DictionaryArray is represented entirely by one ArrayDataRef, **data**.If we do switch to standard iterators, the keys() method would redundant and should go as the iterator is generally better in performance terms.

Reply:
Iterators have a much better chance of becoming SIMD operations, especially if we could annotate the pointer with an alignment.

Reply:
As a compromise, I've dropped the slice based keys method in favour of iterators. It would be nice to see iterators everywhere in the design as indexing is unsafe and inefficient. I've sketched in a sample **NullableIterator** that should use pointers to the buffer data in an ideal design but currently refs the data indirectly. I also sketched in a **FromIterator** implementation so that you could create dictionaries using **collect()**. Again, these concepts could be used globally and IMHO should replace the existing index-based methods.